File: src/battle/order.ts
/**
 * Turn order.
 *
 * Commands for the whole round are chosen first, then everyone acts in one
 * ordered pass. Order is agility plus a small deterministic jitter, with
 * action priority tiers layered on top, so a fast healer is usually first but
 * never reliably so - and a priority move beats raw speed outright.
 *
 * Recomputing per round (rather than a persistent turn queue) means a speed
 * buff applied this round is felt next round, which is easy to reason about.
 */

import { nextFloat, type RngState } from '../core/rng.js';

export interface OrderEntry {
  combatantId: string;
  agility: number;
  priority: number;
  /** Resolved sort key, exposed for debugging and tests. */
  score: number;
}

export interface OrderInput {
  combatantId: string;
  agility: number;
  priority?: number;
}

/** Fraction of agility that the random jitter can swing. */
export const AGILITY_JITTER = 0.15;

export function buildTurnOrder(rng: RngState, inputs: readonly OrderInput[]): OrderEntry[] {
  const entries: OrderEntry[] = inputs.map((input) => {
    const jitter = 1 + (nextFloat(rng) * 2 - 1) * AGILITY_JITTER;
    return {
      combatantId: input.combatantId,
      agility: input.agility,
      priority: input.priority ?? 0,
      score: Math.max(0, input.agility) * jitter,
    };
  });

  entries.sort((a, b) => {
    if (b.priority !== a.priority) return b.priority - a.priority;
    if (b.score !== a.score) return b.score - a.score;
    return a.combatantId < b.combatantId ? -1 : a.combatantId > b.combatantId ? 1 : 0;
  });

  return entries;
}

File: unity/Aetherlight.Core/Battle/Order.cs
using System;
using System.Collections.Generic;

namespace Aetherlight.Battle
{
    /// <summary>
    /// Commands for the whole round are chosen first, then everyone acts in one
    /// ordered pass. Order is agility plus a small jitter, with action priority
    /// tiers on top: a fast healer is usually first but never reliably so, and
    /// a priority move beats raw speed outright.
    ///
    /// Recomputed each round rather than kept as a persistent queue, so a speed
    /// buff applied this round is felt next round.
    /// </summary>
    public sealed class OrderInput
    {
        public string CombatantId = "";
        public double Agility;
        public int Priority;
    }

    public sealed class OrderEntry
    {
        public string CombatantId = "";
        public double Agility;
        public int Priority;
        /// <summary>Resolved sort key, exposed for debugging and tests.</summary>
        public double Score;
    }

    public static class TurnOrder
    {
        /// <summary>Fraction of agility the random jitter can swing.</summary>
        public const double AgilityJitter = 0.15;

        public static List<OrderEntry> Build(ref Core.RngState rng, IReadOnlyList<OrderInput> inputs)
        {
            var entries = new List<OrderEntry>(inputs.Count);
            foreach (var input in inputs)
            {
                double jitter = 1 + (Core.Rng.NextDouble(ref rng) * 2 - 1) * AgilityJitter;
                entries.Add(new OrderEntry
                {
                    CombatantId = input.CombatantId,
                    Agility = input.Agility,
                    Priority = input.Priority,
                    Score = Math.Max(0, input.Agility) * jitter,
                });
            }

            entries.Sort((a, b) =>
            {
                if (b.Priority != a.Priority) return b.Priority.CompareTo(a.Priority);
                if (!b.Score.Equals(a.Score)) return b.Score.CompareTo(a.Score);
                return string.CompareOrdinal(a.CombatantId, b.CombatantId);
            });

            return entries;
        }
    }
}
